backend/SC4SNMP_UI_backend/apply_changes/handle_changes.py

```diff
@@ -0,0 +1,153 @@
+import time
+from abc import abstractmethod, ABC
+from celery import shared_task
+from threading import Lock
+import datetime
+import os
+from kubernetes import client, config
+import yaml
+from kubernetes.client import ApiException
+from SC4SNMP_UI_backend import mongo_client
+from SC4SNMP_UI_backend.apply_changes.kubernetes_job import create_job_object, create_job
+from pymongo import MongoClient
+
+CHANGES_INTERVAL_SECONDS = 300
+MONGO_URI = os.getenv("MONGO_URI")
+JOB_CONFIG_PATH = os.getenv("JOB_CONFIG_PATH", "/config/job_config.yaml")
+JOB_NAMESPACE = os.getenv("JOB_NAMESPACE", "sc4snmp")
+mongo_config_collection = mongo_client.sc4snmp.config_collection
+
+class Handler(ABC):
+    @abstractmethod
+    def set_next(self, handler):
+        pass
+
+    @abstractmethod
+    def handle(self, request):
+        pass
+
+
+class AbstractHandler(Handler):
+    _next_handler: Handler = None
+    def set_next(self, handler: Handler) -> Handler:
+        self._next_handler = handler
+        return handler
+
+    @abstractmethod
+    def handle(self, request: dict):
+        if self._next_handler:
+            return self._next_handler.handle(request)
+        return None
+
+
+class CheckJobHandler(AbstractHandler):
+    def handle(self, request: dict=None):
+        """
+        CheckJobHandler checks whether a new kubernetes job with updated sc4snmp configuration can be run immediately
+         or should it be scheduled for the future.
+
+        :return: pass dictionary with job_delay in seconds to the next handler
+        """
+        record = list(mongo_config_collection.find())[0]
+        last_update = record["previous_job_start_time"]
+        if last_update is None:
+            # If it's the first time that the job is run (last_update is None) then
+            # job delay should be equal to CHANGES_INTERVAL_SECONDS. Update the mongo record with job state accordingly.
+            job_delay = CHANGES_INTERVAL_SECONDS
+            mongo_config_collection.update_one({"_id": record["_id"]},
```

Comment:
Don't we need some kind of `upsert` here? If this is the first record.

Reply:
This record is created [here](https://github.com/splunk/SC4SNMP-UI/blob/fix/add-apply-changes-functionality/backend/SC4SNMP_UI_backend/apply_changes/handle_changes.py#L110-L118) in the constructor of [ApplyChanges](https://github.com/splunk/SC4SNMP-UI/blob/fix/add-apply-changes-functionality/backend/SC4SNMP_UI_backend/apply_changes/handle_changes.py#L102). Code that you marked is part of the handling chain called in the [apply_changes](https://github.com/splunk/SC4SNMP-UI/blob/fix/add-apply-changes-functionality/backend/SC4SNMP_UI_backend/apply_changes/handle_changes.py#L121) method of the `ApplyChanges` changes class. So when it is called, the record already exists but when it is the first time that `apply_changes` is run then `previous_job_start_time` will be `None`, because this is the value that is set in the `ApplyChanges` constructor

Reply:
On the other hand adding `upsert` isn't a big change so I can add it

Reply:
Done